### PR TITLE
#2139: Allow both duration and end fields to be used in API

### LINKED ISSUE #2139: API call adding a timesheet entry in the past without end date is saved with incorrect length
**Describe the bug**
I'm trying to import timesheets from Harvest. I only have the dates and the durations of the logs, so I'm planning to import them by specifying just begin and duration. In the UI this works fine as the end field is automatically populated. However, for past entries entered through the API the end field is null and therefore the timesheet is shown as being active in the UI.

When I edit the timesheet, the duration information is correct and the end field is generated correctly, but it's not really feasible to manually edit all the imported timesheets. 

**To Reproduce**
1. Add a past timesheet entry through the API and leave out the end field altogether, for example with the following curl command:
```
curl -v -X POST "http://localhost:8001/api/timesheets" -H  "accept: application/json" -H "X-AUTH-USER: xxx" -H "X-AUTH-TOKEN: xxxxxx" -H  "Content-Type: application/json" -d "{  \"begin\": \"2020-11-22T01:34:27\",  \"duration\": \"05:00\",  \"project\": 1,  \"activity\": 2,  \"description\": \"string\",  \"tags\": \"string\",  \"fixedRate\": 0,  \"hourlyRate\": 0,  \"user\": 1,  \"exported\": true}"
```
2. Go to the UI. The UI shows the duration of the entry incorrectly due to it being still ”active”.
3. Edit the entry, which fixes the end field and the duration.

**Logfile**
```
[Wed Nov 25 00:04:45.932250 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: SELECT t0.id AS id_1, t0.name AS name_2, t0.color AS color_3 FROM kimai2_tags t0 WHERE t0.name = ? ["string"] []
[Wed Nov 25 00:04:45.952996 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: INSERT INTO kimai2_timesheet (start_time, end_time, timezone, duration, description, rate, internal_rate, fixed_rate, hourly_rate, exported, billable, category, modified_at, `user`, activity_id, project_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) {"1":"2020-11-22 01:34:27","2":null,"3":"Europe/Helsinki","4":18000,"5":"string","6":0,"7":0,"8":0,"9":0,"10":true,"11":true,"12":"work","13":"2020-11-25 02:04:45","14":1,"15":2,"16":1} []
[Wed Nov 25 00:04:45.965500 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: INSERT INTO kimai2_timesheet_tags (timesheet_id, tag_id) VALUES (?, ?) [10,1] []
[Wed Nov 25 00:04:45.971599 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: SELECT k0_.id AS id_0, k0_.start_time AS start_time_1, k0_.end_time AS end_time_2, k0_.timezone AS timezone_3, k0_.duration AS duration_4, k0_.description AS description_5, k0_.rate AS rate_6, k0_.internal_rate AS internal_rate_7, k0_.fixed_rate AS fixed_rate_8, k0_.hourly_rate AS hourly_rate_9, k0_.exported AS exported_10, k0_.billable AS billable_11, k0_.category AS category_12, k0_.modified_at AS modified_at_13, k0_.`user` AS user_14, k0_.activity_id AS activity_id_15, k0_.project_id AS project_id_16 FROM kimai2_timesheet k0_ WHERE k0_.start_time IS NOT NULL AND k0_.end_time IS NULL AND k0_.user = ? ORDER BY k0_.start_time DESC [1] []
[Wed Nov 25 00:04:45.979343 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: SELECT k0_.id AS id_0, k1_.id AS id_1, k1_.name AS name_2, k1_.order_number AS order_number_3, k1_.order_date AS order_date_4, k1_.start AS start_5, k1_.end AS end_6, k1_.timezone AS timezone_7, k1_.comment AS comment_8, k1_.visible AS visible_9, k1_.budget AS budget_10, k1_.time_budget AS time_budget_11, k1_.color AS color_12, k0_.`user` AS user_13, k0_.activity_id AS activity_id_14, k0_.project_id AS project_id_15, k1_.customer_id AS customer_id_16 FROM kimai2_timesheet k0_ LEFT JOIN kimai2_projects k1_ ON k0_.project_id = k1_.id WHERE k0_.id IN (10) [] []
[Wed Nov 25 00:04:45.983053 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: SELECT k0_.id AS id_0, k1_.id AS id_1, k1_.name AS name_2, k1_.number AS number_3, k1_.comment AS comment_4, k1_.visible AS visible_5, k1_.company AS company_6, k1_.vat_id AS vat_id_7, k1_.contact AS contact_8, k1_.address AS address_9, k1_.country AS country_10, k1_.currency AS currency_11, k1_.phone AS phone_12, k1_.fax AS fax_13, k1_.mobile AS mobile_14, k1_.email AS email_15, k1_.homepage AS homepage_16, k1_.timezone AS timezone_17, k1_.budget AS budget_18, k1_.time_budget AS time_budget_19, k1_.color AS color_20, k0_.customer_id AS customer_id_21 FROM kimai2_projects k0_ LEFT JOIN kimai2_customers k1_ ON k0_.customer_id = k1_.id WHERE k0_.id IN (1) [] []
[Wed Nov 25 00:04:45.986165 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: SELECT k0_.id AS id_0, k1_.id AS id_1, k1_.name AS name_2, k1_.comment AS comment_3, k1_.visible AS visible_4, k1_.budget AS budget_5, k1_.time_budget AS time_budget_6, k1_.color AS color_7, k0_.`user` AS user_8, k0_.activity_id AS activity_id_9, k0_.project_id AS project_id_10, k1_.project_id AS project_id_11 FROM kimai2_timesheet k0_ LEFT JOIN kimai2_activities k1_ ON k0_.activity_id = k1_.id WHERE k0_.id IN (10) [] []
[Wed Nov 25 00:04:45.990151 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: SELECT k0_.id AS id_0, k1_.username AS username_1, k1_.username_canonical AS username_canonical_2, k1_.email AS email_3, k1_.email_canonical AS email_canonical_4, k1_.enabled AS enabled_5, k1_.salt AS salt_6, k1_.password AS password_7, k1_.last_login AS last_login_8, k1_.confirmation_token AS confirmation_token_9, k1_.password_requested_at AS password_requested_at_10, k1_.roles AS roles_11, k1_.id AS id_12, k1_.alias AS alias_13, k1_.registration_date AS registration_date_14, k1_.title AS title_15, k1_.avatar AS avatar_16, k1_.api_token AS api_token_17, k1_.auth AS auth_18, k0_.`user` AS user_19, k0_.activity_id AS activity_id_20, k0_.project_id AS project_id_21 FROM kimai2_timesheet k0_ LEFT JOIN kimai2_users k1_ ON k0_.`user` = k1_.`id` WHERE k0_.id IN (10) [] []
[Wed Nov 25 00:04:45.994078 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: SELECT k0_.id AS id_0, k1_.id AS id_1, k1_.name AS name_2, k1_.color AS color_3, k0_.`user` AS user_4, k0_.activity_id AS activity_id_5, k0_.project_id AS project_id_6 FROM kimai2_timesheet k0_ LEFT JOIN kimai2_timesheet_tags k2_ ON k0_.id = k2_.timesheet_id LEFT JOIN kimai2_tags k1_ ON k1_.id = k2_.tag_id WHERE k0_.id IN (10) [] []
[Wed Nov 25 00:04:45.998701 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:45] doctrine.DEBUG: SELECT k0_.id AS id_0, k1_.id AS id_1, k1_.name AS name_2, k1_.value AS value_3, k1_.visible AS visible_4, k0_.`user` AS user_5, k0_.activity_id AS activity_id_6, k0_.project_id AS project_id_7, k1_.timesheet_id AS timesheet_id_8 FROM kimai2_timesheet k0_ LEFT JOIN kimai2_timesheet_meta k1_ ON k0_.id = k1_.timesheet_id WHERE k0_.id IN (10) [] []
[Wed Nov 25 00:04:46.005353 2020] [php7:notice] [pid 133] [client 172.17.0.1:38922] [2020-11-25 00:04:46] doctrine.DEBUG: "COMMIT" [] []
```

**Additional context**
 - Kimai version: 1.11.1 stable
 - PHP version: 7.4.12
 - Database: MySQL

### PR DESCRIPTION
**Pull request by ZeiP**
## Description
I stumbled again on the duration field not working in the API the way I think it should when creating a SailfishOS client for Kimai. Here's an attempt at fixing the issue. Let me know what you think! This should allow using both the duration and end field in the API and in my tests seemed to work perfectly both in API and in the web UI.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] I verified that my code applies to the guidelines (`composer code-check`)
- [ ] I updated the documentation (see [here](https://github.com/kimai/www.kimai.org/tree/master/_documentation))
- [X] I agree that this code is used in Kimai and will be published under the [MIT license](https://github.com/kevinpapst/kimai2/blob/master/LICENSE)

Fixes #2139.

### CODE REVIEW

File: src/Form/TimesheetEditForm.php [hunk 1]
```diff
@@ -202,12 +203,12 @@ function (FormEvent $event) {
                 /** @var Timesheet $data */
                 $data = $event->getData();
                 $duration = $data->getDuration();
-                $end = null;
-                if (null !== $duration) {
+                if (!empty($duration)) {
```
Comment (kevinpapst) [inline src/Form/TimesheetEditForm.php:206]: is this change necessary?

File: src/Form/TimesheetEditForm.php [hunk 2]
```diff
@@ -202,12 +203,12 @@ function (FormEvent $event) {
...
+                    $end = null;
```
Comment (kevinpapst) [inline src/Form/TimesheetEditForm.php:207]: why? the next line sets $end, was that move by accident?

**Review by kevinpapst — CHANGES_REQUESTED** [review]
Thanks for your contribution @ZeiP

Can you have a look at my comments and the the failed tests?

File: src/Form/TimesheetEditForm.php (hunk 1 above)
Comment (ZeiP) [reply src/Form/TimesheetEditForm.php:206]: Yep, although I'm naturally not sure if it's the right thing to do. Basically duration was somewhere reset to 0 if it was missing from the API which caused the data to be saved incorrectly; using empty() requires the value to be both not null and not zero.

File: src/Form/TimesheetEditForm.php (hunk 2 above)
Comment (ZeiP) [reply src/Form/TimesheetEditForm.php:207]: I moved the setEnd() call to inside the if(), so it made sense to move the whole $end variable handling there too – this is to prevent us from overriding the end value if we don't alter it here. But you're absolutely right, this line is meaningless here and should be removed.
